lib/streamlit/elements/file_uploader.py

```diff
@@ -41,6 +48,70 @@
 
 
 class FileUploaderMixin:
+    @overload
+    def file_uploader(
+        self,
+        label: str,
+        type: Optional[Union[str, List[str]]],
+        accept_multiple_files: Literal[True],
+        key: Optional[Key] = None,
+        help: Optional[str] = None,
+        on_change: Optional[WidgetCallback] = None,
+        args: Optional[WidgetArgs] = None,
+        kwargs: Optional[WidgetKwargs] = None,
+        *,
+        disabled: bool = False,
+    ) -> Optional[List[UploadedFile]]:
+        ...
+
+    @overload
+    def file_uploader(
+        self,
+        label: str,
+        type: Optional[Union[str, List[str]]],
+        accept_multiple_files: Literal[False] = False,
+        key: Optional[Key] = None,
+        help: Optional[str] = None,
+        on_change: Optional[WidgetCallback] = None,
+        args: Optional[WidgetArgs] = None,
+        kwargs: Optional[WidgetKwargs] = None,
+        *,
+        disabled: bool = False,
+    ) -> Optional[UploadedFile]:
+        ...
+
+    @overload
+    def file_uploader(
+        self,
+        label: str,
+        *,
+        accept_multiple_files: Literal[True],
+        type: Optional[Union[str, List[str]]] = None,
+        key: Optional[Key] = None,
+        help: Optional[str] = None,
+        on_change: Optional[WidgetCallback] = None,
+        args: Optional[WidgetArgs] = None,
+        kwargs: Optional[WidgetKwargs] = None,
+        disabled: bool = False,
+    ) -> Optional[List[UploadedFile]]:
+        ...
+
+    @overload
+    def file_uploader(
+        self,
+        label: str,
+        *,
+        accept_multiple_files: Literal[False] = False,
+        type: Optional[Union[str, List[str]]] = None,
+        key: Optional[Key] = None,
+        help: Optional[str] = None,
+        on_change: Optional[WidgetCallback] = None,
+        args: Optional[WidgetArgs] = None,
+        kwargs: Optional[WidgetKwargs] = None,
+        disabled: bool = False,
+    ) -> Optional[UploadedFile]:
+        ...
```

Comment:
This may be due to me not being very familiar with the nuances of how `mypy` works, but is there a difference between these two type signatures and the first two?

I would have thought the two signatures are effectively the same because the implementation itself does not enforce that most arguments are passed by keyword.

Reply:
Yes, they are necessary and have difference.

I understand your question, and I actually committed the code without these 2 overloads once (https://github.com/streamlit/streamlit/pull/4188/commits/46a0f596876dfb04019bd020d5553ef72d5f132c).
However, without these overloads, mypy outputs errors below:
```
lib/streamlit/elements/file_uploader.py: note: In member "file_uploader" of class "FileUploaderMixin":
lib/streamlit/elements/file_uploader.py:52:5: error: Overloaded function signatures 1 and 2 overlap with incompatible return types  [misc]
```

This problem was discussed at [this issue](https://github.com/python/mypy/issues/4020#issuecomment-737600893) and its solution had also been introduced at the following comments.

In short, the overloads must cover all variations of argument order, keyword/non-keyword argument combinations and default argument values, and that's why all these overloads are necessary.
For example, without the last 2 overloads, the following types cannot be inferred.
```
st.file_uploader("Single")  # `type` and `accept_multiple_files` are skipped and the default values are used for them
st.file_uploader("Single", accept_multiple_files=False)  # `type` is skipped and the default value is used for it
st.file_uploader("Multi", accept_multiple_files=True)  # `type` is skipped and the default value is used for it
```